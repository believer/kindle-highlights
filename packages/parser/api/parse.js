const lineByLine = require('n-readlines')
const { promisify } = require('util')
const path = require('path')
const fs = require('fs')
const { allowCors } = require('../utils/allowCors')

const writeFile = promisify(fs.writeFile)
const removeFile = promisify(fs.unlink)

const metadata = (line) => {
  const meta = line
    .toString('utf8')
    .match(
      /(page (?<page>\d+) \| )?location (?<location>(\d+|\d+-\d+)) \| added on (?<unparsedDate>.+)/i
    )

  const { page, location, unparsedDate } = meta?.groups ?? {}
  const parsedDate = Date.parse(unparsedDate)
  const date = parsedDate ? new Date(parsedDate).toISOString() : null

  return {
    page,
    location: location.includes('-') ? location.split('-')[0] : location,
    date,
  }
}

const titleAndAuthors = (line) => {
  const titleRow = line
    .toString('utf8')
    .match(/^(?<title>.+)\s\((?<author>.+)\)\s?$/)

  const { title, author } = titleRow?.groups ?? {}

  const titleWithSeries = title.match(/\((?<series>.+)\)/i)

  const { series } = titleWithSeries?.groups ?? {}

  return {
    title: series ? title.replace(/ \((.+)\)/, '').trim() : title,
    series,
    authors: author.split(';').map((author) => {
      if (author.includes(',')) {
        const [lastName, firstName] = author.split(', ')

        return `${firstName.trim()} ${lastName.trim()}`
      }

      return author
    }),
  }
}

const handler = async (req, res) => {
  const { data } = JSON.parse(req.body)
  let line
  const output = []
  const fileName = path.join('/tmp', 'temp.txt')
  const ids = {}

  await writeFile(fileName, data)

  const liner = new lineByLine(fileName)

  while ((line = liner.next())) {
    if (line.toString('utf8').startsWith('====')) {
      const body = []

      // Skip block start line
      // Next line, title and author
      line = liner.next()
      const { title, authors } = titleAndAuthors(line)

      // Next line, meta data
      line = liner.next()
      const { date, page, location } = metadata(line)

      // Skip space between meta and content
      // Next line, content
      line = liner.next()
      line = liner.next()

      // Find content
      while (!line.toString('utf8').startsWith('====')) {
        body.push(line.toString('utf8').trim())
        line = liner.next()
      }

      const id = Buffer.from(body[0]).toString('base64')

      if (body[0] !== '\r' && !ids[id]) {
        output.push({
          authors,
          id,
          body,
          date,
          location,
          page,
          title,
        })

        ids[id] = true
      }
    }
  }

  await removeFile(fileName)

  res.json(output.slice().sort((a, b) => a.title.localeCompare(b.title)))
}

module.exports = allowCors(handler)
