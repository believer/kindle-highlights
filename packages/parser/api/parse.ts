import lineByLine from 'n-readlines'
import { promisify } from 'util'
import path from 'path'
import fs from 'fs'
import { allowCors } from '../utils/allowCors'
import md5 from 'md5'
import { NowRequest, NowResponse } from '@vercel/node'

const writeFile = promisify(fs.writeFile)
const removeFile = promisify(fs.unlink)

interface Metadata {
  page: string
  location: string
  date: string | null
}

export const metadata = (line: Buffer | false): Metadata => {
  const meta = line
    .toString('utf8')
    .match(
      /on (page (?<page>\w+))?( \| )?(location (?<location>(\d+-\d+|\d+)))?( \| )?(added on (?<unparsedDate>.+))?/i
    )

  const { page = '', location = '', unparsedDate } = meta?.groups ?? {}
  const parsedDate = Date.parse(unparsedDate)
  const date = parsedDate ? new Date(parsedDate).toISOString() : null

  return {
    page,
    location: location.includes('-') ? location.split('-')[0] : location,
    date,
  }
}

interface TitleAuthor {
  title: string
  series: string | null
  authors: string[]
}

export const titleAndAuthors = (line: Buffer | false): TitleAuthor => {
  try {
    if (line.toString('utf8').startsWith('- Your')) {
      throw new Error('Highlight row as title')
    }

    const titleRow = line
      .toString('utf8')
      .match(
        /^(?<title>[^\(]+)( ?\((?<series>[^\)]+)\))?( ?\((?<author>[^\)]+))?/
      )

    const { title, author, series } = titleRow?.groups ?? {}
    const parsedAuthor = series && !author ? series : author

    const titleWithoutSubtitle = title.includes(':')
      ? title.replace(/:.+/, '')
      : title

    return {
      title: series
        ? titleWithoutSubtitle.replace(/ \((.+)\)/, '').trim()
        : titleWithoutSubtitle.trim(),
      series: series && author ? series : null,
      authors: parsedAuthor
        ? parsedAuthor.split(';').map((author: string) => {
            if (author.includes(',')) {
              const [lastName, firstName] = author.split(', ')

              return `${firstName.trim()} ${lastName.trim()}`
            }

            return author
          })
        : [],
    }
  } catch (e) {
    console.error(e, line.toString('utf8'))

    return {
      title: '',
      authors: [],
      series: null,
    }
  }
}

interface IssueProps {
  bodyLength: number
  containsSimilar: boolean
}

enum Issue {
  similar = 'similar',
  short = 'shortContent',
}

export const findIssues = ({
  bodyLength,
  containsSimilar,
}: IssueProps): Issue[] => {
  const issues = []

  if (containsSimilar) {
    issues.push(Issue.similar)
  }

  if (bodyLength <= 10) {
    issues.push(Issue.short)
  }

  return issues
}

const isValidBody = ({
  body,
  ids,
  id,
}: {
  body: string[]
  ids: Record<string, boolean>
  id: string
}): boolean => body[0] !== '\r' && body[0] !== '' && !ids[id]

export const handler = async (req: NowRequest, res: NowResponse) => {
  const { data } = JSON.parse(req.body)
  let line: Buffer | false
  const output = []
  const fileName = path.join('/tmp', 'temp.txt')
  const ids: Record<string, boolean> = {}

  await writeFile(fileName, data)

  const liner = new lineByLine(fileName)

  while ((line = liner.next())) {
    const body = []

    // Skip block start line
    // Next line, title and author
    const { title, authors, series } = titleAndAuthors(line)

    // Next line, meta data
    if (!line.toString('utf8').startsWith('- Your')) {
      line = liner.next()
    }

    const { date, page, location } = metadata(line)

    // Skip space between meta and content
    // Next line, content
    line = liner.next()
    line = liner.next()

    // Find content
    while (!line.toString('utf8').startsWith('====')) {
      body.push(line.toString('utf8').trim())
      line = liner.next()
    }

    const content = body.join('\n')
    const id = md5(content)

    // If a similar exist for the same page, location and title
    // remove it and only keep the last one
    const containsSimilar: number = output.findIndex(
      (highlight) =>
        highlight.title === title &&
        highlight.page === page &&
        highlight.location === location
    )

    if (containsSimilar !== -1) {
      output.splice(containsSimilar, 1)
    }

    if (isValidBody({ body, ids, id })) {
      output.push({
        authors,
        id,
        content,
        date,
        series,
        location,
        page,
        title,
        bookId: md5(title),
        issues: findIssues({
          containsSimilar: containsSimilar > -1,
          bodyLength: content.length,
        }),
      })

      ids[id] = true
    }
  }

  await removeFile(fileName)

  res.json(output.slice().sort((a, b) => a.title.localeCompare(b.title)))
}

export default allowCors(handler)
