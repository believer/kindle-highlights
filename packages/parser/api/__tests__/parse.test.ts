import { metadata, handler, titleAndAuthors } from '../parse'

describe('#metadata', () => {
  test('parses page, location and date from line', () => {
    const line = Buffer.from(
      '- Your Highlight on page 94 | Location 1388-1389 | Added on Monday, April 8, 2019 9:52:25 PM'
    )

    expect(metadata(line)).toEqual({
      page: '94',
      location: '1388',
      date: '2019-04-08T19:52:25.000Z',
    })
  })

  test('handles metadata with single location', () => {
    const line = Buffer.from(
      '- Your Highlight on page 94 | Location 1388 | Added on Monday, April 8, 2019 9:52:25 PM'
    )

    expect(metadata(line)).toEqual({
      page: '94',
      location: '1388',
      date: '2019-04-08T19:52:25.000Z',
    })
  })

  test('handles metadata with roman page number', () => {
    const line = Buffer.from(
      '- Your Highlight on page xvii | Location 138-141 | Added on Monday, April 12, 2021 9:53:39 PM'
    )

    expect(metadata(line)).toEqual({
      page: 'xvii',
      location: '138',
      date: '2021-04-12T19:53:39.000Z',
    })
  })

  test('handles metadata without page', () => {
    const line = Buffer.from(
      '- Your Highlight on Location 76-78 | Added on Saturday, October 31, 2020 11:52:03 PM'
    )

    expect(metadata(line)).toEqual({
      page: '',
      location: '76',
      date: '2020-10-31T22:52:03.000Z',
    })
  })

  test('handles metadata without location', () => {
    const line = Buffer.from(
      '- Your Highlight on page 76 | Added on Saturday, October 31, 2020 11:52:03 PM'
    )

    expect(metadata(line)).toEqual({
      page: '76',
      location: '',
      date: '2020-10-31T22:52:03.000Z',
    })
  })

  test('handles metadata without date', () => {
    const line = Buffer.from('- Your Highlight on page 76 | Location 138-14')

    expect(metadata(line)).toEqual({
      page: '76',
      location: '138',
      date: null,
    })
  })
})

describe('#titleAndAuthors', () => {
  test('handles parsing title and author', () => {
    const line = Buffer.from('Make Time (Knapp, Jake)')

    expect(titleAndAuthors(line)).toEqual({
      title: 'Make Time',
      authors: ['Jake Knapp'],
      series: null,
    })
  })

  test('handles parsing title and multiple authors', () => {
    const line = Buffer.from(
      'The Daily Stoic (Holiday, Ryan;Hanselman, Stephen)'
    )

    expect(titleAndAuthors(line)).toEqual({
      title: 'The Daily Stoic',
      authors: ['Ryan Holiday', 'Stephen Hanselman'],
      series: null,
    })
  })

  test('handles parsing title and multiple authors without lastname, firstname', () => {
    const line = Buffer.from(
      'Barn som bråkar : att hantera känslostarka barn i vardagen (Bo Hejlskov;Tina Wiman)'
    )

    expect(titleAndAuthors(line)).toEqual({
      title: 'Barn som bråkar',
      authors: ['Bo Hejlskov', 'Tina Wiman'],
      series: null,
    })
  })

  test('handles parsing title, author and series', () => {
    const line = Buffer.from(
      'Astrophysics for People in a Hurry (Astrophysics for People in a Hurry Series) (de Grasse Tyson, Neil)'
    )

    expect(titleAndAuthors(line)).toEqual({
      title: 'Astrophysics for People in a Hurry',
      authors: ['Neil de Grasse Tyson'],
      series: 'Astrophysics for People in a Hurry Series',
    })
  })
})

describe('#handler', () => {
  test('can parse two entries', async () => {
    const data = `Samlingsvolym (Lars Kepler)
- Your Bookmark on Location 26219 | Added on Tuesday, August 9, 2016 7:14:43 PM


==========
Deep Work: Rules for Focused Success in a Distracted World (Newport, Cal)
- Your Highlight on page 50 | Location 601-602 | Added on Tuesday, March 19, 2019 5:28:00 PM

of one, a lonely voice issuing ex
==========
The Subtle Art of Not Giving a F*ck (Manson, Mark)
- Your Highlight on page 177 | Location 2338-2340 | Added on Monday, May 27, 2019 10:26:05 PM

Rather, a healthy relationship is when two people solve their own problems in order to feel good about each other.
==========
The Subtle Art of Not Giving a F*ck (Manson, Mark)
- Your Highlight on page 181 | Location 2384-2386 | Added on Monday, May 27, 2019 10:30:32 PM

It’s not about giving a fuck about everything your partner gives a fuck about; it’s about giving a fuck about your partner regardless of the fucks he or she gives. That’s unconditional love, baby.
==========
Barn som bråkar : att hantera känslostarka barn i vardagen (Bo Hejlskov;Tina Wiman)
- Your Highlight on page 110 | Location 1684-1699 | Added on Sunday, August 16, 2020 9:57:33 PM

annat ändras? Detta kan visserligen betyda att vi föräldrar ibland måste göra upp med våra inre föreställningar om hur saker måste vara. Men om vi inte ändrar förutsättningarna kan vi inte heller räkna med att bråken blir färre. Många bråk uppstår när barnen – eller vi vuxna – är hungriga eller trötta. Det kan vara innan maten eller när barnen kommer hem från skolan eller vi från jobbet. Det betyder att det ofta går att förutse när på dagen det kommer att bli bråk och att vi då kan planera in fungerande aktiviteter, mellanmål eller tid för vila. Tänk på att även vi vuxna behöver vila och äta på strategiska tider för att orka förebygga bråk. Håll utkik efter de situationer där barnen inte bråkar, om det finns några. Vad är det som är annorlunda då? Försök att använda det även i andra situationer. Gör mer av de aktiviteter som fungerar. Mer tid i fungerande situationer betyder mindre tid med bråk. Se till att förbereda med bra strukturer vid gemensamma utflykter. Om familjen tillsammans ska bestämma efterhand vad som ska hända kommer det definitivt att bli bråk. Gör i stället ett schema för hur dagen ska se ut. Bestäm i förväg var ni ska äta och när det bär av hemåt igen. Och låt gärna barnen vara med och bestämma, strukturen kan göras i god tid tillsammans med dem. Men det är viktigt att komma ihåg att det alltid är föräldrarna som har sista ordet. Se inte syskonbråk som ett föräldramisslyckande. Försök i stället att se det som situationer som dels ska lösas här och nu, dels förebyggas på sikt. Då är det lättare att hålla sig lugn. Vi har sagt det förut, men det tål att upprepas – syskon bråkar med varandra. Det är något mycket vardagligt.
==========
`

    const req: any = {
      body: JSON.stringify({ data }),
    }

    const res: any = { json: jest.fn() }

    await handler(req, res)

    expect(res.json.mock.calls[0][0]).toMatchSnapshot()
  })

  test('only include the latest if multiple clippings', async () => {
    const data = `Ego is the Enemy (The Way, the Enemy and the Key) (Holiday, Ryan)
- Your Highlight on page 193 | Location 2227-2227 | Added on Tuesday, March 23, 2021 10:59:33 PM

“He who fears death will never do anything worthy of a living man,” Seneca once
==========
Ego is the Enemy (The Way, the Enemy and the Key) (Holiday, Ryan)
- Your Highlight on page 193 | Location 2227-2227 | Added on Tuesday, March 23, 2021 10:59:39 PM

“He who fears death will never do anything worthy of a living man,” Seneca once said.
==========
Ego is the Enemy (The Way, the Enemy and the Key) (Holiday, Ryan)
- Your Highlight on page 193 | Location 2227-2228 | Added on Tuesday, March 23, 2021 10:59:48 PM

“He who fears death will never do anything worthy of a living man,” Seneca once said. Alter that: He who will do anything to avoid failure will almost
==========
Ego is the Enemy (The Way, the Enemy and the Key) (Holiday, Ryan)
- Your Highlight on page 193 | Location 2227-2228 | Added on Tuesday, March 23, 2021 10:59:53 PM

“He who fears death will never do anything worthy of a living man,” Seneca once said. Alter that: He who will do anything to avoid failure will almost certainly do something worthy of a failure.
==========`

    const req: any = {
      body: JSON.stringify({ data }),
    }

    const res: any = { json: jest.fn() }

    await handler(req, res)

    expect(res.json.mock.calls[0][0]).toMatchSnapshot()
  })
})
