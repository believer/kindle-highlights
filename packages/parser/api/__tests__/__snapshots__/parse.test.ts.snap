// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#handler can parse two entries 1`] = `
Array [
  Object {
    "authors": Array [
      "Bo Hejlskov",
      "Tina Wiman",
    ],
    "content": "annat ändras? Detta kan visserligen betyda att vi föräldrar ibland måste göra upp med våra inre föreställningar om hur saker måste vara. Men om vi inte ändrar förutsättningarna kan vi inte heller räkna med att bråken blir färre. Många bråk uppstår när barnen – eller vi vuxna – är hungriga eller trötta. Det kan vara innan maten eller när barnen kommer hem från skolan eller vi från jobbet. Det betyder att det ofta går att förutse när på dagen det kommer att bli bråk och att vi då kan planera in fungerande aktiviteter, mellanmål eller tid för vila. Tänk på att även vi vuxna behöver vila och äta på strategiska tider för att orka förebygga bråk. Håll utkik efter de situationer där barnen inte bråkar, om det finns några. Vad är det som är annorlunda då? Försök att använda det även i andra situationer. Gör mer av de aktiviteter som fungerar. Mer tid i fungerande situationer betyder mindre tid med bråk. Se till att förbereda med bra strukturer vid gemensamma utflykter. Om familjen tillsammans ska bestämma efterhand vad som ska hända kommer det definitivt att bli bråk. Gör i stället ett schema för hur dagen ska se ut. Bestäm i förväg var ni ska äta och när det bär av hemåt igen. Och låt gärna barnen vara med och bestämma, strukturen kan göras i god tid tillsammans med dem. Men det är viktigt att komma ihåg att det alltid är föräldrarna som har sista ordet. Se inte syskonbråk som ett föräldramisslyckande. Försök i stället att se det som situationer som dels ska lösas här och nu, dels förebyggas på sikt. Då är det lättare att hålla sig lugn. Vi har sagt det förut, men det tål att upprepas – syskon bråkar med varandra. Det är något mycket vardagligt.",
    "date": "2020-08-16T19:57:33.000Z",
    "id": "bc5e71897bc82dc32d582e8d3a3ebb7a",
    "issues": Array [],
    "location": "1684",
    "page": "110",
    "title": "Barn som bråkar",
  },
  Object {
    "authors": Array [
      "Cal Newport",
    ],
    "content": "of one, a lonely voice issuing ex",
    "date": "2019-03-19T16:28:00.000Z",
    "id": "709e3cff6f9446ea2ca786a8ac21bf56",
    "issues": Array [],
    "location": "601",
    "page": "50",
    "title": "Deep Work",
  },
  Object {
    "authors": Array [
      "Mark Manson",
    ],
    "content": "Rather, a healthy relationship is when two people solve their own problems in order to feel good about each other.",
    "date": "2019-05-27T20:26:05.000Z",
    "id": "be6265c5b34192df7308388621b77609",
    "issues": Array [],
    "location": "2338",
    "page": "177",
    "title": "The Subtle Art of Not Giving a F*ck",
  },
  Object {
    "authors": Array [
      "Mark Manson",
    ],
    "content": "It’s not about giving a fuck about everything your partner gives a fuck about; it’s about giving a fuck about your partner regardless of the fucks he or she gives. That’s unconditional love, baby.",
    "date": "2019-05-27T20:30:32.000Z",
    "id": "12065bcac905f31654df1b7298147170",
    "issues": Array [],
    "location": "2384",
    "page": "181",
    "title": "The Subtle Art of Not Giving a F*ck",
  },
]
`;

exports[`#handler only include the latest if multiple clippings 1`] = `
Array [
  Object {
    "authors": Array [
      "Ryan Holiday",
    ],
    "content": "“He who fears death will never do anything worthy of a living man,” Seneca once said. Alter that: He who will do anything to avoid failure will almost certainly do something worthy of a failure.",
    "date": "2021-03-23T21:59:53.000Z",
    "id": "556f6bc53437f85b0d3420a8deb9402e",
    "issues": Array [
      "similar",
    ],
    "location": "2227",
    "page": "193",
    "title": "Ego is the Enemy",
  },
]
`;
